ygen/enumgen.go

```diff
@@ -74,61 +74,186 @@ func newEnumGenState() *enumGenState {
+	definingModName := genutil.ParentModulePrettyName(i)
+	// As per a typedef that includes an enumeration, there is a many to one
+	// relationship between leaves and an identity value, therefore, we want to
+	// reuse the existing name for the identity enumeration if one exists.
+	return fmt.Sprintf("%s/%s", definingModName, i.Name)
+}
+
+// enumeratedTypedefKey calculates a unique string key for the input typedef
+// *yang.Entry that has an underlying enumerated type (e.g., identityref or
+// enumeration). It also returns the defining module and type name components
+// of the identity for use in the name generation, if needed.
+func (s *enumGenState) enumeratedTypedefKey(e *yang.Entry, noUnderscores bool) (string, string, string, error) {
+	typeName := e.Type.Name
+
+	// Handle the case whereby we have been handed an enumeration that is within a
+	// union. We need to synthesise the name of the type here such that it is based on
+	// type name, plus the fact that it is an enumeration.
+	if e.Type.Kind == yang.Yunion {
+		enumTypes := util.EnumeratedUnionTypes(e.Type.Type)
+
+		switch len(enumTypes) {
+		case 1:
+			// We specifically say that this is an enumeration within the leaf.
+			if noUnderscores {
+				typeName = fmt.Sprintf("%sEnum", enumTypes[0].Name)
 			} else {
-				var err error
-				enumName, err = s.resolveTypedefEnumeratedName(e, noUnderscores)
-				if err != nil {
-					return nil, err
-				}
+				typeName = fmt.Sprintf("%s_Enum", enumTypes[0].Name)
 			}
+		case 0:
+			return "", "", "", fmt.Errorf("enumerated type had an empty union within it, path: %v, type: %v, enumerated: %v", e.Path(), e.Type, enumTypes)
+		default:
+			return "", "", "", fmt.Errorf("multiple enumerated types within a single enumeration not supported, path: %v, type: %v, enumerated: %v", e.Path(), e.Type, enumTypes)
+		}
+	}
+	if e.Node == nil {
+		return "", "", "", fmt.Errorf("nil Node in enum type %s", e.Name)
+	}
 
-			en = &yangEnum{
-				name: enumName,
-				entry: &yang.Entry{
-					Name: e.Name,
-					Type: &yang.YangType{
-						Name: e.Type.Name,
-						Kind: yang.Yenum,
-						Enum: t.Enum,
-					},
-					Annotation: map[string]interface{}{"valuePrefix": util.SchemaPathNoChoiceCase(e)},
-				},
-			}
+	definingModName := genutil.ParentModulePrettyName(e.Node)
+	// Since there can be many leaves that refer to the same typedef, then we do not generate
+	// a name for each of them, but rather use a common name, we use the non-CamelCase lookup
+	// as this is unique, whereas post-camelisation, we may have name clashes. Since a typedef
+	// does not have a 'path' in Goyang, so we synthesise one using the form
+	// module-name/typedef-name.
+	return fmt.Sprintf("%s/%s", definingModName, typeName), definingModName, typeName, nil
+}
+
+// enumLeafKey calculates a unique string key for the input enum leaf.  If
+// compressPaths is true, it also returns the compress name of the entry for
+// use in name generation, if needed.
```

Comment:
Shall we highlight here that this covers leaves of type `enumeration` only? I believe that this is the case based on the name, but it might be useful to be explicit.